components/epd_driver/epd_driver.c

```diff
@@ -301,39 +310,278 @@ inline uint32_t min(uint32_t x, uint32_t y) { return x < y ? x : y; }
 
 void epd_draw_hline(int x, int y, int length, uint8_t color,
                     uint8_t *framebuffer) {
-  if (y < 0 || y >= EPD_HEIGHT) {
-    return;
-  }
   for (int i = 0; i < length; i++) {
     int xx = x + i;
-    if (xx < 0 || xx >= EPD_WIDTH) {
-      continue;
-    }
-    uint8_t *buf_ptr = &framebuffer[y * EPD_WIDTH / 2 + xx / 2];
-    if (xx % 2) {
-      *buf_ptr = (*buf_ptr & 0x0F) | (color & 0xF0);
-    } else {
-      *buf_ptr = (*buf_ptr & 0xF0) | (color >> 4);
-    }
+    epd_draw_pixel(xx, y, color, framebuffer);
   }
 }
 
 void epd_draw_vline(int x, int y, int length, uint8_t color,
                     uint8_t *framebuffer) {
+  for (int i = 0; i < length; i++) {
+    int yy = y + i;
+    epd_draw_pixel(x, yy, color, framebuffer);
+  }
+}
+
+void epd_draw_pixel(int x, int y, uint8_t color,
+                    uint8_t *framebuffer) {
   if (x < 0 || x >= EPD_WIDTH) {
     return;
   }
-  for (int i = 0; i < length; i++) {
-    int yy = y + i;
-    if (yy < 0 || yy >= EPD_HEIGHT) {
-      return;
+  if (y < 0 || y >= EPD_HEIGHT) {
+    return;
+  }
+  uint8_t *buf_ptr = &framebuffer[y * EPD_WIDTH / 2 + x / 2];
+  if (x % 2) {
+    *buf_ptr = (*buf_ptr & 0x0F) | (color & 0xF0);
+  } else {
+    *buf_ptr = (*buf_ptr & 0xF0) | (color >> 4);
+  }
+}
+
+void epd_draw_circle(int x0, int y0, int r, uint8_t color,
+                    uint8_t *framebuffer) {
+  int f = 1 - r;
+  int ddF_x = 1;
+  int ddF_y = -2 * r;
+  int x = 0;
+  int y = r;
+
+  epd_draw_pixel(x0, y0 + r, color, framebuffer);
+  epd_draw_pixel(x0, y0 - r, color, framebuffer);
+  epd_draw_pixel(x0 + r, y0, color, framebuffer);
+  epd_draw_pixel(x0 - r, y0, color, framebuffer);
+
+  while (x < y) {
+    if (f >= 0) {
+      y--;
+      ddF_y += 2;
+      f += ddF_y;
+    }
+    x++;
+    ddF_x += 2;
+    f += ddF_x;
+
+    epd_draw_pixel(x0 + x, y0 + y, color, framebuffer);
+    epd_draw_pixel(x0 - x, y0 + y, color, framebuffer);
+    epd_draw_pixel(x0 + x, y0 - y, color, framebuffer);
+    epd_draw_pixel(x0 - x, y0 - y, color, framebuffer);
+    epd_draw_pixel(x0 + y, y0 + x, color, framebuffer);
+    epd_draw_pixel(x0 - y, y0 + x, color, framebuffer);
+    epd_draw_pixel(x0 + y, y0 - x, color, framebuffer);
+    epd_draw_pixel(x0 - y, y0 - x, color, framebuffer);
+  }
+}
+
+void epd_fill_circle(int x0, int y0, int r, uint8_t color,
+                    uint8_t *framebuffer) {
+  epd_draw_vline(x0, y0 - r, 2 * r + 1, color, framebuffer);
+  epd_fill_circle_helper(x0, y0, r, 3, 0, color, framebuffer);
+}
+
+
+void epd_fill_circle_helper(int16_t x0, int16_t y0, int16_t r,
```

Comment:
Is there a specific reason for the int16_t type? If not, I'd rather keep the types consistent and just use int.

Reply:
yep, let's use int

Reply:
done